### PR TITLE
perf(collaboration): collaboration support for VSCode

### PR DESCRIPTION
**Pull request by SLdragon**
### Most of the files are not related to our PR, just some lint changes, wenyu have already tried to fix this issue, so you can just ignore these files.

Our changes are mostly in below files:
packages/fx-core/src/plugins/solution/fx-solution/solution.ts 
packages/vscode-extension/src/commandsTreeViewProvider.ts
packages/vscode-extension/src/envTree.ts
packages/vscode-extension/src/extension.ts
packages/vscode-extension/src/handlers.ts

Demo picture:
image

### CODE REVIEW

**Review by hund030 — APPROVED** [review]
Seems only code style changed in function plugin.

File: packages/vscode-extension/src/envTree.ts [hunk 1]
```diff
@@ -44,9 +52,45 @@ export async function registerEnvTreeHandler(): Promise<Result<Void, FxError>> {
           isCustom: false,
           description:
             item === activeEnv ? StringResources.vsc.commandsTreeViewProvider.acitve : "",
+          expanded: true,
```
[the remark this answers is not in the document]
Comment (SLdragon) [reply packages/vscode-extension/src/envTree.ts:55]: Sure, will discuss with Yanliang and Jaeyong

[the remark this answers is not in the document]
Comment (SLdragon) [reply packages/vscode-extension/src/envTree.ts:55]: If user cannot access the environment, we will show "No permission to list collaborators" on the tree view with a warning icon

**Review by tecton — APPROVED** [review]
SDK part looks good.

File: packages/vscode-extension/CHANGELOG.md [hunk 2]
```diff
@@ -3,14 +3,16 @@
 ## 2.6.0 - Sep 06, 2021
 
 Incremental version for Teams Toolkit with multiple bugs fixed and the following updates：
- 
+
```
Comment (adashen) [inline packages/vscode-extension/CHANGELOG.md:6]: why we have change in this file?

File: packages/vscode-extension/src/handlers.ts [hunk 3]
```diff
@@ -795,6 +797,143 @@ export async function activateEnvironment(env: string): Promise<Result<Void, FxE
...
+    window.showInformationMessage(
+      `Added account: '${inputs.email}'' to the environment '${env}' as a collaborator`
+    );
+
+    updateCollaboratorList(env);
+  } catch (e) {
+    result = wrapError(e);
+  }
+
+  await processResult(eventName, result);
```
Comment (adashen) [inline packages/vscode-extension/src/handlers.ts:827]: where do we send the success/error telemetry for grant?

File: packages/vscode-extension/CHANGELOG.md (hunk 2 above)
Comment (SLdragon) [reply packages/vscode-extension/CHANGELOG.md:6]: These are lint changes, after we merge some conflict, there are so may lint changes in out PR, wenyu will help to solve these lint changes issue in the future

File: packages/vscode-extension/src/handlers.ts (hunk 3 above)
Comment (SLdragon) [reply packages/vscode-extension/src/handlers.ts:827]: processResult function will help show error message dialog and send telemetry

**Review by 1yefuwang1 — APPROVED** [review]

File: packages/vscode-extension/src/commandsTreeViewProvider.ts [hunk 4]
```diff
@@ -496,13 +496,37 @@ export class CommandsTreeViewProvider implements vscode.TreeDataProvider<TreeVie
               );
               this.disposableMap.set(subTreeItem.commandId, disposable);
             }
+
+            let tooltip: string | vscode.MarkdownString = treeItem.label;
```
Comment (HuihuiWu-Microsoft) [inline packages/vscode-extension/src/commandsTreeViewProvider.ts:500]: If subTreeItem.tooltip equals undefined and treeItem.label exists, then the tooltip of subTreeItem will becomes treeItem.label?

File: packages/vscode-extension/src/envTree.ts [hunk 5]
```diff
@@ -44,9 +52,45 @@ export async function registerEnvTreeHandler(): Promise<Result<Void, FxError>> {
...
+          ? "environmentWithPermission"
+          : "environment";
+        if (isRemoteCollaborateEnabled()) {
+          userList = await listCollaborator(env);
+        }
+      } else {
+        userList = [
+          {
+            commandId: `fx-extension.listcollaborator.${env}`,
+            label: "Login M365 account to view all collaborators",
```
Comment (HuihuiWu-Microsoft) [inline packages/vscode-extension/src/envTree.ts:86]: Better move to strings.json if the string is available to users for later localization.

File: packages/vscode-extension/src/envTree.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SLdragon) [reply packages/vscode-extension/src/envTree.ts]: Discussed with Yanliang and Jaeyong, we reached an agreement that only expand active environment list by default. Code is updated, thanks!

File: packages/vscode-extension/src/handlers.ts [hunk 6]
```diff
@@ -795,6 +797,143 @@ export async function activateEnvironment(env: string): Promise<Result<Void, FxE
...
+          isMarkdown: false,
+        },
+        parent: "fx-extension.environment." + env,
+      };
+    });
+    if (!result || result.length === 0) {
+      result = [
+        {
+          commandId: `fx-extension.listcollaborator.${env}`,
+          label: "No permission to list collaborators",
```
Comment (HuihuiWu-Microsoft) [inline packages/vscode-extension/src/handlers.ts:864]: Same here

File: packages/vscode-extension/src/commandsTreeViewProvider.ts (hunk 4 above)
Comment (SLdragon) [reply packages/vscode-extension/src/commandsTreeViewProvider.ts:500]: Yes, that's true

File: packages/vscode-extension/src/envTree.ts (hunk 5 above)
Comment (SLdragon) [reply packages/vscode-extension/src/envTree.ts:86]: OK, thanks for your tip, will update

**Review by dooriya — APPROVED** [review]

File: packages/vscode-extension/src/envTree.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dooriya) [reply packages/vscode-extension/src/envTree.ts]: Thanks Renlong!

File: packages/vscode-extension/src/envTree.ts (hunk 5 above)
Comment (SLdragon) [reply packages/vscode-extension/src/envTree.ts]: Updated

File: packages/vscode-extension/src/handlers.ts (hunk 6 above)
Comment (SLdragon) [reply packages/vscode-extension/src/handlers.ts]: Updated

**Review by KennethBWSong — APPROVED** [review]

**Review by qinezh — APPROVED** [review]

**Review by HuihuiWu-Microsoft — APPROVED** [review]
LGTM for extension part

File: packages/vscode-extension/src/commandsTreeViewProvider.ts (hunk 4 above)
Comment (SLdragon) [reply packages/vscode-extension/src/commandsTreeViewProvider.ts]: Updated this logic, thanks!

**Review by nliu-ms — APPROVED** [review]

**Pull request merged**